merkle-tree/indexed/src/lib.rs

```diff
@@ -171,58 +234,129 @@ where
-                let patched_element = IndexedElement::<I> {
-                    value: BigUint::from_bytes_be(&changelog_element.value),
-                    index: changelog_element.index,
-                    next_index: changelog_element.next_index,
-                };
-                // Removing the value:
-                // Writing data costs CU thus we just overwrite the index
-                // with an impossible value so that it cannot be found.
-                changelog_element.index = max_usize
-                    .try_into()
-                    .map_err(|_| IndexedMerkleTreeError::IntegerOverflow)?;
-                // Only use changelog event values, since these originate from an account -> can be trusted
-                Ok(Some((patched_element, changelog_element.next_value)))
+            })
+            .collect();
+
+        let mut new_low_element = None;
+
+        for next_indexed_changelog_index in next_indexed_changelog_indices {
+            let changelog_entry = &mut self.indexed_changelog[next_indexed_changelog_index];
+
+            let next_element_value = BigUint::from_bytes_be(&changelog_entry.element.next_value);
+            if next_element_value < new_element.value {
+                // If the next element is lower than the current element, it means
+                // that it should become the low element.
+                //
+                // Save it and break the loop.
+                new_low_element = Some((changelog_entry.element.next_value, next_element_value));
+                break;
             }
-            None => Ok(None),
+
+            // Patch the changelog index.
+            *changelog_index = changelog_entry.changelog_index;
+
+            // Patch the `next_index` of `new_element`.
+            new_element.next_index = changelog_entry.element.next_index;
+
+            // Patch the element.
+            low_element.update_from_raw_element(&changelog_entry.element);
+            // Patch the next value.
+            *low_element_next_value = BigUint::from_bytes_be(&changelog_entry.element.next_value);
+            // Patch the proof.
+            low_leaf_proof.clone_from(&changelog_entry.proof);
         }
+
+        // If we found a new low element.
+        if let Some((new_low_element_raw, new_low_element)) = new_low_element {
+            // Find a changelog entry which represents the insertion of
+            // new low element.
+            //
+            // PANICS: At this point, we are absolutely sure that such entry exists.
+            // It it doesn't, then we really did something wrong with the
+            // implementation.
+            let new_low_element_changelog_index = self
+                .indexed_changelog
+                .iter_from(indexed_changelog_index)?
+                .skip(1)
+                .position(|changelog_entry| changelog_entry.element.value == new_low_element_raw)
+                .unwrap()
```

Comment:
instead of looking for the index can't we use ``next_indexed_changelog_index`` from line 273?